File: code/python/index1.py
#!/usr/bin/env python

import csv
import json
import sys
import xapian

def index(datapath, dbpath):
    """Index a set of data.

    :param datapath The path to the file containing the incoming data.
    :param dbpath The path to the database to index to.

    This will create the database if it doesn't already exist.

    """

    # Create or open the database we're going to be writing to. 
    db = xapian.WritableDatabase(dbpath, xapian.DB_CREATE_OR_OPEN)
    termgenerator = xapian.TermGenerator()
    termgenerator.set_stemmer(xapian.Stem("en"))

    for fields in parse_csv_file(datapath):
        # fields is a dictionary keyed by field name.  We're just going to use
        # the id_NUMBER, TITLE and DESCRIPTION fields to start with.
        doc = xapian.Document()

        termgenerator.set_document(doc)
        termgenerator.index_text(fields['TITLE'], 'S')
        termgenerator.index_text(fields['DESCRIPTION'], 'XD')

        termgenerator.index_text(fields['TITLE'])
        termgenerator.increase_termpos()
        termgenerator.index_text(fields['DESCRIPTION'])

        doc.set_data(json.dumps(fields))
        db.add(doc)

def parse_csv_file(datapath):
    """Parse a CSV file, yielding dictionaries of the fields.

    """
    fd = open(datapath)
    reader = csv.reader(fd)

    # Read the first 
    titles = reader.next()
    for row in reader:
        yield dict(zip(titles, row))
    fd.close()


if __name__ == '__main__':
    datapath = sys.argv[1]
    dbpath = sys.argv[2]
    index(datapath, dbpath)

File: code/java/index1.java
package code.java;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;
import java.util.ArrayList;
import java.util.logging.Level;
import java.util.logging.Logger;
import org.xapian.Document;
import org.xapian.Stem;
import org.xapian.TermGenerator;
import org.xapian.WritableDatabase;
import org.xapian.XapianConstants;
import org.xapian.XapianJNI;

import code.java.support;

public class index1 {

    // Command line args - datapath dbpath
    public static void main(String[] args) {
        if (args.length < 2) {
            System.out.println("Insufficient number of arguments (should be datapath dbpath)");
            return;
        }
        index(args[0], args[1]);
    }

    // Start of example code.
    public static void index(String datapath, String dbpath) {
        // Create or open the database we're going to be writing to.
        WritableDatabase db = new WritableDatabase(dbpath, XapianConstants.DB_CREATE_OR_OPEN);
        // Set up a TermGenerator that we'll use in indexing.
        TermGenerator termGenerator = new TermGenerator();
        termGenerator.setStemmer(new Stem("en"));

        // Parsing the CSV input file
        Scanner csvScanner = null;

        try {
            File csv = new File(datapath);
            csvScanner = new Scanner(csv);
        } catch (FileNotFoundException ex) {
            Logger.getLogger(index1.class.getName()).log(Level.SEVERE, null, ex);
        }

        // Ignoring first line (contains descriptors)
        csvScanner.nextLine();

        while (csvScanner.hasNextLine()) {
            String currentLine = csvScanner.nextLine();

            /* Parsing each line for identifier, title, and description */
            ArrayList<String> parsedCSV = support.parseCsvLine(currentLine);
            // Identifier is the first comma seperated value (according to CSV file)
            String identifier = parsedCSV.get(0);

            // Title is third comma seperated value
            String title = parsedCSV.get(2);

            // Description is ninth comma sperated value
            String description = parsedCSV.get(8);

            /* Finished Parsing line */

            // We make a document and tell the term generator to use this.
            Document doc = new Document();
            termGenerator.setDocument(doc);

            // Index each field with a suitable prefix.
            termGenerator.indexText(title, 1, "S");
            termGenerator.indexText(description, 1, "XD");

            // Index fields without prefixes for general search.
            termGenerator.indexText(title);
            termGenerator.increaseTermpos();
            termGenerator.indexText(description);

            // Store all fields for display purposes
            doc.setData(currentLine);
            doc.addValue(0, title);

            // We use the identifier to ensure each object ends up in the
            // database only once no matter how many times we run the
            // indexer.
            String idterm = "Q"+identifier;
            doc.addBooleanTerm(idterm);
            db.replaceDocument(idterm, doc);
        }

        // Commit to write documents to disk
        db.commit();
        csvScanner.close();
    }
    // End of example code.

}
